compute-hypothetical-feature.mjs

```diff
@@ -0,0 +1,143 @@
+            // it has no usage in the caniuse data.
+            return String(Number.MAX_SAFE_INTEGER);
+        }
+        if (version == '15.2-15.3') {
+            return '15.2';
+        }
+        return version;
+    }
+    return compare(normalize(a), normalize(b), cmp);
+}
+
+// timestamp is number of seconds since epoch
+function caniuseDate(timestamp) {
+  return new Date(1000 * timestamp);
+}
+
+// from (inclusive) and to (exclusive) are Date objects.
+// TODO: this isn't quite right, date 
+function* iterateMonths(from, to) {
+    const iter = new Date(from);
+    while (iter < to) {
+        yield new Date(iter);
+        // Add one calendar month.
+        const nextMonth = (iter.getUTCMonth() + 1) % 12;
+        if (nextMonth == 0) {
+            iter.setUTCFullYear(iter.getUTCFullYear() + 1);
+        }
+        iter.setUTCMonth(nextMonth);
+    }
+}
+
+function monthString(date) {
+    return date.toISOString().substr(0, 10);
+}
+
+// Find the usage of a browser at a given date. The exact date is probably not
+// found in |historicalBrowserData|, so use the latest entry at or before
+// |date|. If there is none, return the very latest entry. The caller needs to
+// check that the data is fresh enough before using it.
+function getHistoricalUsage(browser, date) {
+    const snapshots = historicalBrowserData[browser];
+    // TODO: binary search would be O(log(n)) instead of O(n).
+    const index = snapshots.findLastIndex((s) => caniuseDate(s.timestamp) <= date);
+    // Index could be -1, and we return the latest snapshot in this case.
+    return snapshots.at(index);
+}
+
+const months = Array.from(iterateMonths(Date.parse('2020-01-01T00:00Z'), endDate));
+
+function getVersionAfter(browser, date) {
+    for (const release of caniuse.agents[browser].version_list) {
+        if (!release.release_date) {
+            continue;
+        }
+        const releaseDate = caniuseDate(release.release_date);
+        if (releaseDate >= date) {
+            return release.version;
+        }
+    }
+    return null;
+}
+
+for (const startIndex of months.keys()) {
+    const launchDate = months[startIndex];
+    // Assume that a hypothetical feature is enabled in all browser engines at
+    // |launchDate| and is available in all following browser releases. Step
+    // forward one month at a time and compute availability based on the
+    // historical usage stats in |historicalBrowserData|.
+
+    // Identify the version of each browser where the feature was first available.
+    // This is done up front to make it simpler to determine if each release has
+    // the feature based on the version, not the release date.
+    const launchVersions = {};
+    for (const browser of browsers) {
+        launchVersions[browser] = getVersionAfter(browser, launchDate);
+    }
+
+    console.log(monthString(launchDate));
+    for (let index = startIndex; index < months.length; index++) {
```

Comment:
Yes, this is O(n²) in the number of months. Doesn't really matter, but an alternative way to think about the problem is starting at today, and working backwards until we reach the point where a feature released at that past time would now have 95% availability. Then we couldn't tell if the time to 95% availability has been shrinking over time, but I think all we'll care about in the end is our expectation for how long it will take today.

The usage data itself is a much bigger problem right now though.